api/src/main/java/org/apache/iceberg/expressions/CountAggregate.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.expressions;
+
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.StructLike;
+
+public class CountAggregate<T> extends BoundAggregate<T, Long> {
+  protected CountAggregate(Operation op, BoundTerm<T> term) {
+    super(op, term);
+  }
+
+  @Override
+  public Long eval(StructLike struct) {
+    return countFor(struct);
+  }
+
+  @Override
+  public Long eval(DataFile file) {
+    return countFor(file);
+  }
+
+  protected Long countFor(StructLike row) {
+    throw new UnsupportedOperationException(
+        this.getClass().getName() + " does not implement countFor(StructLike)");
+  }
+
+  protected Long countFor(DataFile file) {
+    throw new UnsupportedOperationException(
+        this.getClass().getName() + " does not implement countFor(DataFile)");
+  }
+
+  @Override
+  public Aggregator<Long> newAggregator() {
+    return new CountAggregator<>(this);
+  }
+
+  private static class CountAggregator<T> extends NullSafeAggregator<T, Long> {
+    private Long count = 0L;
+
+    CountAggregator(BoundAggregate<T, Long> aggregate) {
+      super(aggregate);
+    }
+
+    @Override
+    protected void update(Long value) {
+      if (value != null) {
```

Comment:
For all cases, this should not be called if the value is null. Handling null values is the purpose of `NullSafeAggregator`.